File: packages/pvm/src/host-functions/accumulate/eject.ts
import { encodeFixedLength, getServiceRequestValue } from '@pbnjam/codec'
import { bytesToHex, hexToBytes, logger } from '@pbnjam/core'
import type { HostFunctionResult } from '@pbnjam/types'
import { ACCUMULATE_FUNCTIONS, RESULT_CODES } from '../../config'
import {
  type AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
} from './base'

/**
 * EJECT accumulation host function (Ω_J)
 *
 * Ejects/removes service account
 *
 * Gray Paper Specification:
 * - Function ID: 21 (eject)
 * - Gas Cost: 10
 * - Parameters: registers[7-8] = d, o
 *   - d: service account ID to eject
 *   - o: hash offset in memory (32 bytes)
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read hash from memory at offset o (32 bytes)
 * 2. Get service account d from accounts dictionary
 * 3. Check if service account exists and is not the current service
 * 4. Verify the hash matches the service's code hash
 * 5. Check if the service has exactly 2 items and the request exists
 * 6. Check if the ejection period has expired (y < t - Cexpungeperiod)
 * 7. Remove the service account and transfer its balance to current service
 */
export class EjectHostFunction extends BaseAccumulateHostFunction {
  readonly functionId = ACCUMULATE_FUNCTIONS.EJECT
  readonly name = 'eject'
  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const { registers, ram, implications, timeslot } = context
    // Extract parameters from registers
    const [serviceIdToEject, hashOffset] = registers.slice(7, 9)

    // Log all input parameters
    context.log('EJECT host function invoked', {
      serviceIdToEject: serviceIdToEject.toString(),
      hashOffset: hashOffset.toString(),
      timeslot: timeslot.toString(),
      currentServiceId: implications[0].id.toString(),
      expungePeriod: context.expungePeriod.toString(),
    })

    // Read hash from memory (32 bytes)
    // Gray Paper line 851-854: h = memory[o:32] when Nrange(o,32) ⊆ readable(memory), error otherwise
    const [hashData, faultAddress] = ram.readOctets(hashOffset, 32n)
    // Gray Paper line 862: panic when h = error, registers[7] unchanged
    if (faultAddress || !hashData) {
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }

    // Get the current implications context
    const [imX] = implications

    // Get service account d from accounts dictionary
    // Gray Paper: d ≠ imX.id ∧ d ∈ keys(imX.state.ps_accounts)
    if (serviceIdToEject === imX.id) {
      this.setAccumulateError(registers, 'WHO')
      return {
        resultCode: null, // continue execution
      }
    }

    const serviceAccount = imX.state.accounts.get(serviceIdToEject)
    if (!serviceAccount) {
      this.setAccumulateError(registers, 'WHO')
      return {
        resultCode: null, // continue execution
      }
    }

    // Verify the hash matches the service's code hash
    // Gray Paper: d.sa_codehash ≠ encode[32]{imX.id}
    // Use encodeFixedLength for proper Gray Paper encoding
    const [encodeError, expectedCodeHash] = encodeFixedLength(imX.id, 32n)
    if (encodeError) {
      logger.error('[EjectHostFunction] Failed to encode service ID', {
        serviceId: imX.id.toString(),
        error: encodeError.message,
      })
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }
    const serviceCodeHash = hexToBytes(serviceAccount.codehash)
    if (!this.arraysEqual(serviceCodeHash, expectedCodeHash)) {
      logger.warning('[EjectHostFunction] Codehash mismatch', {
        serviceIdToEject: serviceIdToEject.toString(),
        currentServiceId: imX.id.toString(),
        expectedCodeHash: bytesToHex(expectedCodeHash),
        actualCodeHash: serviceAccount.codehash,
      })
      this.setAccumulateError(registers, 'WHO')
      return {
        resultCode: null, // continue execution
      }
    }

    // Calculate length: max(81, d.sa_octets) - 81
    const l = Math.max(81, Number(serviceAccount.octets)) - 81

    // Check if the service has exactly 2 items and the request exists
    // Gray Paper: d.sa_items ≠ 2 ∨ (h, l) ∉ d.sa_requests
    if (Number(serviceAccount.items) !== 2) {
      this.setAccumulateError(registers, 'HUH')
      return {
        resultCode: null, // continue execution
      }
    }

    // Get request using nested map structure: requests[hash][length]
    const hashHex = bytesToHex(hashData)
    const requestValue = getServiceRequestValue(serviceAccount, serviceIdToEject, hashHex, BigInt(l));

    // const requestMap = serviceAccount.requests.get(hashHex)
    if (!requestValue) {
      logger.warning('[EjectHostFunction] Request map not found for hash', {
        hashHex,
        availableHashes: Object.keys(serviceAccount.rawCshKeyvals),
      })
      this.setAccumulateError(registers, 'HUH')
      return {
        resultCode: null, // continue execution
      }
    }


    // Check if the ejection period has expired
    // Gray Paper: d.sa_requests[h, l] = [x, y], y < t - Cexpungeperiod
    // For test vectors, Cexpungeperiod = 32 (as per README)
    // For production, Cexpungeperiod = 19200 (Gray Paper constant)
    const expungePeriod = context.expungePeriod
    const threshold = timeslot - expungePeriod

    if (requestValue.length < 2 || requestValue[1] >= timeslot - expungePeriod) {
      logger.warning('[EjectHostFunction] Expunge period check failed', {
        requestLength: requestValue.length,
        y: requestValue[1]?.toString(),
        timeslot: timeslot.toString(),
        expungePeriod: expungePeriod.toString(),
        threshold: threshold.toString(),
      })
      this.setAccumulateError(registers, 'HUH')
      return {
        resultCode: null, // continue execution
      }
    }

    // Transfer balance to current service and remove the ejected service
    // Gray Paper: imX'.state.ps_accounts = imX.state.ps_accounts \ {d} ∪ {imX.id: s'}
    // where s' = imX.self except s'.sa_balance = imX.self.sa_balance + d.sa_balance
    const currentService = imX.state.accounts.get(imX.id)

    if (!currentService) {
      logger.error(
        '[EjectHostFunction] Current service account not found in state',
        {
          currentServiceId: imX.id.toString(),
          availableAccountIds: Array.from(imX.state.accounts.keys()).map((id) =>
            id.toString(),
          ),
        },
      )
      this.setAccumulateError(registers, 'WHO')
      return {
        resultCode: null, // continue execution
      }
    }

    currentService.balance += serviceAccount.balance

    // Remove the ejected service account
    imX.state.accounts.delete(serviceIdToEject)

    // Set success result
    this.setAccumulateSuccess(registers)
    return {
      resultCode: null, // continue execution
    }
  }

  private arraysEqual(a: Uint8Array, b: Uint8Array): boolean {
    if (a.length !== b.length) return false
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) return false
    }
    return true
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/eject.rs
//! EJECT accumulation host function (Ω_J). Gray Paper: function ID 21.
//! r7 = service ID to eject (d), r8 = hash offset (o). 1:1 with AS eject.ts.

use crate::config::FUNC_EJECT;
use crate::host_functions::accumulate::base::{self, codes};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};
use crate::codec::{
    decode_request_timeslots, encode_fixed_length, get_request_value,
};

const HASH_LEN: u32 = 32;

pub struct EjectHostFunction;

impl HostFunction for EjectHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_EJECT
    }
    fn name(&self) -> &'static str {
        "eject"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let service_id_to_eject = context.registers[7];
        let hash_offset = context.registers[8];

        let _log_service_id = context.service_id.unwrap_or(0);
        crate::host_log!(
            "[hostfn] EJECT host function invoked serviceIdToEject={} hashOffset={} currentServiceId={}",
            service_id_to_eject, hash_offset, _log_service_id
        );

        let read_result = context.ram.read_octets(hash_offset as u32, HASH_LEN);
        if read_result.fault_address != 0 || read_result.data.is_none() {
            crate::host_log!(
                "[hostfn] eject PANIC: hash read fault (offset={}, fault_address={})",
                hash_offset, read_result.fault_address
            );
            return HostFunctionResult::panic();
        }
        let hash_data = read_result.data.unwrap();
        if hash_data.len() != HASH_LEN as usize {
            crate::host_log!(
                "[hostfn] eject PANIC: hash length mismatch (got {}, expected {})",
                hash_data.len(),
                HASH_LEN as usize
            );
            return HostFunctionResult::panic();
        }
        let mut hash_bytes = [0u8; 32];
        hash_bytes.copy_from_slice(&hash_data);

        let current_service_id = match context.service_id {
            Some(id) => id,
            None => {
                base::set_accumulate_error(context.registers, codes::WHO);
                return HostFunctionResult::continue_execution();
            }
        };
        let accounts = match context.accounts.as_deref_mut() {
            Some(a) => a,
            None => {
                base::set_accumulate_error(context.registers, codes::WHO);
                return HostFunctionResult::continue_execution();
            }
        };

        if service_id_to_eject == current_service_id {
            base::set_accumulate_error(context.registers, codes::WHO);
            return HostFunctionResult::continue_execution();
        }

        let target = match accounts.get(&service_id_to_eject) {
            Some(t) => t,
            None => {
                base::set_accumulate_error(context.registers, codes::WHO);
                return HostFunctionResult::continue_execution();
            }
        };

        let expected_code_hash = encode_fixed_length(current_service_id, 32);
        if target.codehash.as_slice() != expected_code_hash.as_slice() {
            base::set_accumulate_error(context.registers, codes::WHO);
            return HostFunctionResult::continue_execution();
        }

        let l = 81u64.max(target.octets) - 81;

        if target.items != 2 {
            base::set_accumulate_error(context.registers, codes::HUH);
            return HostFunctionResult::continue_execution();
        }

        let request_value = match get_request_value(
            target,
            service_id_to_eject as u32,
            &hash_bytes,
            l,
        ) {
            Some(v) => v,
            None => {
                base::set_accumulate_error(context.registers, codes::HUH);
                return HostFunctionResult::continue_execution();
            }
        };

        let timeslots = match decode_request_timeslots(&request_value) {
            Some(t) => t,
            None => {
                base::set_accumulate_error(context.registers, codes::HUH);
                return HostFunctionResult::continue_execution();
            }
        };

        let timeslot = context.timeslot.unwrap_or(0);
        let expunge_period = context.expunge_period.unwrap_or(0);
        if timeslots.len() < 2 {
            base::set_accumulate_error(context.registers, codes::HUH);
            return HostFunctionResult::continue_execution();
        }
        let y = u64::from(timeslots[1]);
        if y + expunge_period >= timeslot {
            base::set_accumulate_error(context.registers, codes::HUH);
            return HostFunctionResult::continue_execution();
        }

        let balance_to_transfer = target.balance;
        accounts.remove(&service_id_to_eject);
        // Resolve current service account (imX) for balance update; from context.service_account or accounts.
        let current_service = match &mut context.service_account {
            Some(s) => s,
            None => match accounts.get_mut(&current_service_id) {
                Some(s) => s,
                None => {
                    base::set_accumulate_error(context.registers, codes::WHO);
                    return HostFunctionResult::continue_execution();
                }
            },
        };
        current_service.balance += balance_to_transfer;

        base::set_accumulate_success(context.registers, codes::OK);
        crate::host_log!("[host-calls] [{}] EJECT({}, {}) <- OK", _log_service_id, service_id_to_eject, hash_offset);
        HostFunctionResult::continue_execution()
    }
}
